# Final Project: Witch Trials in Europe - 01 Data Cleaning

This dataset contains data of more than 43,000 people tried for witchcraft across 21 European countries over a period of five-and-a-half centuries. It was compiled by economists Peter T. Leeson and Jake Russ. ([On GitHub](https://github.com/JakeRuss/witch-trials); [On Kaggle](https://www.kaggle.com/datasets/michaelbryantds/witch-trials?resource=download)) It contains 12 columns and 10940 entries.

Pros: 
- Relatively small and clean. 
- Witch trials is such a significant historical part of early modern history that there is no lack of research around it. Many of the observations and analysis based on the data can either be confirmed or explained by relevant historical events.

Cons: 
- The problem is that, because the data is so old, many of the entries are null or incomplete. 
- It should be noted that the concept of nation in early modern Europe can be very different from the contemporary understanding of nation-states. For example, Germany did not exist until the 1800s, and for the majority of the timeline were part of the Holy Roman Empire (composed of the areas of present-day Germany, Switzerland and Austria).

What the data can tell us:
- general trend across years and countries.
- the peak of the witch trials. 

What the data can't tell us:
- the first and the last witch trial case.
- the explanation for such a mass hysteria.
- the demography of those who were prosecuted, which, according to wikipedia, are mostly women.

Questions to explore:
- Q1: What time period (century/decade/year) witnessed the peak of witch trials?
- Q2: Which country has the most witch trial cases in total?
- Q3: Which country has the least witch trial cases in total?
- Q4: How did the trend spread across Europe?

## import modules and dataset

In [1]:
import pandas as pd
import altair as alt

In [2]:
trials_original = pd.read_csv('trials.csv', dtype={'year': str})

In [3]:
trials = trials_original.copy()

In [4]:
trials.sample(20)

year  decade  century  tried  deaths               city  \
6206   1432    1430     1400      1     1.0           Konstanz   
9002   1597    1590     1500      1     NaN           Budingen   
9409   1662    1660     1600      1     NaN  Offenbach am Main   
5847   1691    1690     1600      1     NaN                NaN   
9471   1579    1570     1500      1     NaN                NaN   
4072   1649    1640     1600      1     0.0                NaN   
10591  1605    1600     1600      1     1.0             Bruges   
4496   1649    1640     1600      3     NaN                NaN   
6160   1607    1600     1600     12     NaN                NaN   
9952   1753    1750     1700      1     NaN                NaN   
2072   1546    1540     1500      1     0.0                NaN   
5856   1697    1690     1600      1     0.0                NaN   
10466  1645    1640     1600      1     1.0                NaN   
8460   1608    1600     1600      1     0.0             Geneva   
239     NaN    1650     1600     28    12.0                NaN   
789    1591    1590     1500      2     NaN        Eggenfelden   
10100  1636    1630     1600      1     NaN                NaN   
9156   1633    1630     1600      1     NaN          Ortenberg   
4021   1649    1640     1600      1     0.0                NaN   
8537   1620    1620     1600      1     0.0             Geneva   

             gadm.adm2          gadm.adm1       gadm.adm0        lon  \
6206          Freiburg  Baden-Wurttemberg         Germany   9.173238   
9002         Darmstadt             Hessen         Germany   9.111380   
9409         Darmstadt             Hessen         Germany   8.776080   
5847         Edinburgh           Scotland  United Kingdom        NaN   
9471        Shropshire            England  United Kingdom        NaN   
4072         Edinburgh           Scotland  United Kingdom        NaN   
10591  West-Vlaanderen         Vlaanderen         Belgium   3.224700   
4496               NaN           Scotland  United Kingdom        NaN   
6160         Magdeburg     Sachsen-Anhalt         Germany        NaN   
9952          Borgholm             Kalmar          Sweden        NaN   
2072             Namur           Wallonie         Belgium        NaN   
5856      Renfrewshire           Scotland  United Kingdom        NaN   
10466          Suffolk            England  United Kingdom        NaN   
8460            Geneve             Geneve     Switzerland   6.142300   
239    Prattigau/Davos         Graubunden     Switzerland        NaN   
789      Mittelfranken             Bayern         Germany  12.757500   
10100              NaN    Vastra Gotaland          Sweden        NaN   
9156         Darmstadt             Hessen         Germany   9.049020   
4021      East Lothian           Scotland  United Kingdom        NaN   
8537            Geneve             Geneve     Switzerland   6.142300   

            lat          record.source  
6206   47.67795      Kieckhefer (1976)  
9002   50.28630           Niess (1982)  
9409   50.09560           Niess (1982)  
5847        NaN  Goodare et al. (2003)  
9471        NaN       Notestein (1968)  
4072        NaN  Goodare et al. (2003)  
10591  51.20940      Vanysacker (1988)  
4496        NaN  Goodare et al. (2003)  
6160        NaN         Kauertz (2001)  
9952        NaN          Sörlin (1999)  
2072        NaN        Brouette (1953)  
5856        NaN  Goodare et al. (2003)  
10466       NaN        Valletta (2000)  
8460   46.19840          Monter (1976)  
239         NaN           Bader (1945)  
789    48.40510       Behringer (1987)  
10100       NaN          Sörlin (1999)  
9156   50.35560           Niess (1982)  
4021        NaN  Goodare et al. (2003)  
8537   46.19840          Monter (1976)

In [5]:
trials.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10940 entries, 0 to 10939
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   year           10009 non-null  object 
 1   decade         10940 non-null  int64  
 2   century        10940 non-null  int64  
 3   tried          10940 non-null  int64  
 4   deaths         7114 non-null   float64
 5   city           5727 non-null   object 
 6   gadm.adm2      9893 non-null   object 
 7   gadm.adm1      10781 non-null  object 
 8   gadm.adm0      10940 non-null  object 
 9   lon            5137 non-null   float64
 10  lat            5137 non-null   float64
 11  record.source  10940 non-null  object 
dtypes: float64(3), int64(3), object(6)
memory usage: 1.0+ MB


There are 10940 rows in total, that means only 5 columns (`decade`, `century`, `tried`, `gadm.adm0` and `record.source`) contain no empty rows.

## Data Cleaning

### Convert float to int

The first thing I noticed is that `decade`, `century` and `tried` are int64 but `year` and `deaths` are float64. That is because int type doesn't allow missing/NaN values. So I'm going to convert the latter two into Int64, an extension types implemented within pandas that allows nullable integers (Note that dtype must be `Int64` and not `int64` (first 'i' must be capitalized)).

In [6]:
# trials = trials.astype({'year':'Int64', 'deaths':'Int64'})
# trials.sample(20)

However, upon this conversion, the NaN was converted into `<NA>`. So maybe the better way is to bring in the year column in `str` in `pd.read_csv()`

### Duplicates

In [7]:
trials_duplicates = trials[trials.duplicated()]
trials_duplicates

year  decade  century  tried  deaths       city      gadm.adm2  \
854    1597    1590     1500      1     NaN   Augsburg  Mittelfranken   
898    1605    1600     1600      1     NaN   Augsburg  Mittelfranken   
980    1612    1610     1600      1     NaN     Munich  Mittelfranken   
1023   1615    1610     1600      1     NaN     Munich  Mittelfranken   
1132   1629    1620     1600      1     NaN     Dachau  Mittelfranken   
...     ...     ...      ...    ...     ...        ...            ...   
10683  1569    1560     1500      1     0.0  Enkhuizen      Enkhuizen   
10695  1594    1590     1500      1     0.0    Haarlem        Haarlem   
10697  1595    1590     1500      2     0.0    Haarlem        Haarlem   
10735  1635    1630     1600      1     0.0    Haarlem        Haarlem   
10826  1585    1580     1500      1     0.0   Schiedam       Schiedam   

           gadm.adm1    gadm.adm0      lon      lat     record.source  
854           Bayern      Germany  10.8978  48.3705  Behringer (1987)  
898           Bayern      Germany  10.8978  48.3705  Behringer (1987)  
980           Bayern      Germany  11.5820  48.1351  Behringer (1987)  
1023          Bayern      Germany  11.5820  48.1351  Behringer (1987)  
1132          Bayern      Germany  11.4339  48.2630  Behringer (1987)  
...              ...          ...      ...      ...               ...  
10683  Noord-Holland  Netherlands      NaN      NaN     Waardt (1991)  
10695  Noord-Holland  Netherlands      NaN      NaN     Waardt (1991)  
10697  Noord-Holland  Netherlands      NaN      NaN     Waardt (1991)  
10735  Noord-Holland  Netherlands      NaN      NaN     Waardt (1991)  
10826   Zuid-Holland  Netherlands      NaN      NaN     Waardt (1991)  

[3884 rows x 12 columns]

There are 3884 duplicated rows. However, it is possible that when certain columns are NaN, pandas will treat the rows as duplicates. As the data is quite old, we can only assume that each row is a new entry. 

### Export a clean csv file

In [8]:
trials.to_csv('./export/trials_clean.csv', index=False)